crates/kumod/src/smtp_server.rs

```diff
@@ -2264,6 +2280,177 @@ impl SmtpServerSession {
+                }
+                size = self.socket.as_mut().unwrap().read(&mut data) => {
+                    match size {
+                        Err(err) => {
+                            tracing::trace!("error reading: {err:#}");
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Diagnostic {
+                                    level: Level::ERROR,
+                                    message: format!("error reading: {err:#}"),
+                                },
+                                when: Utc::now(),
+                            });
+                            return Err(anyhow!("error reading proxy protocol header: {err:#}"));
+                        }
+                        Ok(size) if size == 0 => {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Diagnostic {
+                                    level: Level::ERROR,
+                                    message: "Peer Disconnected".to_string(),
+                                },
+                                when: Utc::now(),
+                            });
+                            return Err(anyhow!("peer disconnected while reading proxy protocol header"));
+                        }
+                        Ok(size) => {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Read(data[0..size].to_vec()),
+                                when: Utc::now(),
+                            });
+                            self.read_buffer.extend_from_slice(&data[0..size]);
+                        }
+                    }
+                }
+                _ = self.shutdown.shutting_down() => {
+                    return Err(anyhow!("shutting down"));
+                }
+            };
+            let header = HeaderResult::parse(&self.read_buffer);
+            if header.is_complete() {
+                break header;
+            }
+        };
+
+        let mut addr: Option<IpAddr> = None;
+        let mut port: Option<u16> = None;
+        let mut dest_addr: Option<IpAddr> = None;
+        let mut dest_port: Option<u16> = None;
+        match header {
+            HeaderResult::V1(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v1::Addresses::Tcp4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
```

Comment:
I think you can do something like this to get a `SocketAddr`:

```suggestion
                        addr.replace((addresses.source_address, addresses.source_port).into());
```

https://doc.rust-lang.org/beta/std/net/enum.SocketAddr.html#impl-From%3C(I,+u16)%3E-for-SocketAddr